crates/noirc_frontend/src/hir_def/types.rs

```diff
@@ -1155,12 +1226,204 @@ impl Type {
+                }
+            }
+
+            (TypeExpression::Constant(x), TypeExpression::Constant(y)) => {
+                if x == y {
+                    Ok(())
+                } else {
+                    Err(SpanKind::None)
+                }
+            }
+
+            (
+                TypeExpression::BinaryOperation(self_lhs, self_op, self_rhs),
+                TypeExpression::BinaryOperation(other_lhs, other_op, other_rhs),
+            ) => {
+                if self_op != other_op {
+                    return Err(SpanKind::None);
+                }
+
+                let mut result = self_lhs.try_unify(other_lhs, span);
+                result = result.and_then(|()| self_rhs.try_unify(other_rhs, span));
+
+                // If we got an error and our operator is commutative, try swapping the ordering of
+                // the two arguments to see if we succeed. Note that this may still fail in cases
+                // like (N + 1 + M) = (1 + M + N) as this would require multiple nested swaps
+                // instead of just one at a time.
+                if result.is_err() && self_op.commutative() {
+                    // No `and_then` here, we want to discard the previous failure.
+                    result = self_lhs.try_unify(other_rhs, span);
+                    result = result.and_then(|()| self_rhs.try_unify(other_lhs, span));
+                }
+
+                result
+            }
+
+            (TypeExpression::Constant(_), TypeExpression::BinaryOperation(_, _, _))
+            | (TypeExpression::BinaryOperation(_, _, _), TypeExpression::Constant(_)) => {
+                // In the future we could try to evaluate both expressions to see if they
+                // simplify down to the same expression. For now we just fail.
+                Err(SpanKind::None)
+            }
+
+            (TypeExpression::NamedGeneric(..), _) | (_, TypeExpression::NamedGeneric(..)) => {
+                Err(SpanKind::None)
+            }
+        }
+    }
+
+    // This could be more robust, but we'd quickly get into dangerous territory with questions of
+    // whether we should group repeated additions into multiplications, re-arrange nested terms,
+    // etc.
+    pub fn simplify(&self) -> TypeExpression {
+        use TypeExpression::*;
+        match self {
+            Constant(_) | TypeVariable(_) | NamedGeneric(..) => self.clone(),
+            BinaryOperation(lhs, op, rhs) => {
+                let lhs = lhs.simplify();
+                let rhs = rhs.simplify();
+
+                match (lhs, rhs) {
+                    (Constant(lhs), Constant(rhs)) => Constant(op.function()(lhs, rhs)),
+                    (lhs, rhs) => BinaryOperation(Rc::new(lhs), *op, Rc::new(rhs)),
+                }
+            }
+        }
+    }
+}
+
+impl BinaryTypeOperator {
+    /// Return the actual rust numeric function associated with this operator
+    fn function(self) -> fn(u64, u64) -> u64 {
```

Comment:
I think we should ensure that no overflow/underflow can happen here

Reply:
The default operators should panic on overflow/underflow in debug, I can change them to `|a, b| a.checked_add(b).expect("overflow occurred during type checking")` to have them panic in release as well.

As for the comptime versus numeric generics comment, I'll take a step back and consider the design of the two again since ideally we should only have one. Originally it was practical limitations that caused the split. Comptime would be the ideal of the two but since they are only evaluated once all functions are inlined during SSA we could not evaluate them during type checking - hence the need for a more restricted version.

If we could evaluate comptime in the frontend during e.g. monomorphisation or a new pass, it may be possible to delay some type checks until this step as well.

Reply:
Looks like `checked_add` is only available via libraries. I've actually changed it to be explicitly wrapping instead to match the behavior runtime integers have in noir. Let me know what you think.

Reply:
The issue I see with wrapping is that the type is implied because generics can be used to define array lengths, and we use u64 for array lengths. So it is really not clear how it will wrap if you are a user. But if checking for overflow is not easily do-able, then I guess wrapping is OK.
I will add an issue for this instead.